File: 55-JumpGame/55-JumpGame.py
# Last updated: 5/6/2025, 11:10:19 pm
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n=len(nums)
        dp=[False]*(n)

        dp[0]=True

        for idx in range(n):
            if not dp[idx]:
                continue
            for jump_len in range(1,nums[idx]+1):
                if idx+jump_len>=n:
                    dp[n-1]=True
                    return dp[n-1]
                
                else:
                    dp[idx+jump_len]=True
        print(dp)
        return dp[n-1]

File: 55-JumpGame/55-JumpGame.cpp
// Last updated: 5/6/2025, 11:10:13 pm
class Solution {
public:
    bool canJump(vector<int>& nums) {
        //it shows at max what index can I reach.
        //initially I can only reach index 0, hence reach = 0
        int reach = 0; 
    
        for(int idx = 0; idx < nums.size(); idx++) {
            //at every index I'll check if my reach was atleast able to 
            //reach that particular index.
            
            //reach >= idx -> great, carry on. Otherwise, 
            if(reach < idx) return false;
            
            //now as you can reach this index, it's time to update your reach
            //as at every index, you're getting a new jump length.
            reach = max(reach, idx + nums[idx]);
        }
        
        //this means that you reached till the end of the array, wohooo!! 
        return true;
        
    }
};